source/server/listener_manager_impl.cc

```diff
@@ -694,20 +706,49 @@ void ListenerManagerImpl::onListenerWarmed(ListenerImpl& listener) {
   updateWarmingActiveGauges();
 }
 
-void ListenerManagerImpl::drainFilterChains(ListenerImplPtr&& listener) {
+void ListenerManagerImpl::inPlaceFilterChainUpdate(ListenerImpl& listener) {
+  auto existing_active_listener = getListenerByName(active_listeners_, listener.name());
+  auto existing_warming_listener = getListenerByName(warming_listeners_, listener.name());
+  ASSERT(existing_warming_listener != warming_listeners_.end());
+  ASSERT(*existing_warming_listener != nullptr);
+
+  (*existing_warming_listener)->debugLog("execute in place filter chain update");
+  if (existing_active_listener != active_listeners_.end()) {
+    ASSERT(*existing_active_listener != nullptr);
+    // The warmed listener should be added first so that the worker will accept new connections
+    // when it stops listening on the old listener.
+    for (const auto& worker : workers_) {
+      addListenerToWorker(*worker, listener.listenerTag(), listener, nullptr);
+    }
+    // Finish active_listeners_ transformation before calling `drainListener` as it depends on their
+    // state.
+    auto previous_listener = std::move(*existing_active_listener);
+    *existing_active_listener = std::move(*existing_warming_listener);
+    drainFilterChains(std::move(previous_listener), **existing_active_listener);
+  } else {
+    for (const auto& worker : workers_) {
+      addListenerToWorker(*worker, absl::nullopt, listener, nullptr);
+    }
+    active_listeners_.emplace_back(std::move(*existing_warming_listener));
+  }
+
+  warming_listeners_.erase(existing_warming_listener);
+  updateWarmingActiveGauges();
+}
+
+void ListenerManagerImpl::drainFilterChains(ListenerImplPtr&& listener,
+                                            ListenerImpl& new_listener) {
```

Comment:
I'm not quite sure what these parameters are, old listener and new listener? Something else? Clarify names?

Reply:
will change to old and new